src/main/java/org/dataone/hashstore/hashfs/HashFileStore.java

```diff
@@ -220,4 +247,244 @@ protected HashAddress put(InputStream object, String pid, String additionalAlgor
+                if (additionalAlgorithm != null) {
+                    extraAlgo.update(buffer, 0, bytesRead);
+                }
+            }
+        } catch (IOException ioe) {
+            logHashFileStore.error(
+                    "HashFileStore.writeToTmpFileAndGenerateChecksums - IOException encountered (os.flush/close or write related): "
+                            + ioe.getMessage());
+            throw ioe;
+        } finally {
+            os.flush();
+            os.close();
+        }
+
+        logHashFileStore.info("HashFileStore.writeToTmpFileAndGenerateChecksums - Object has been written to tmpFile");
+        String md5Digest = DatatypeConverter.printHexBinary(md5.digest()).toLowerCase();
+        String sha1Digest = DatatypeConverter.printHexBinary(sha1.digest()).toLowerCase();
+        String sha256Digest = DatatypeConverter.printHexBinary(sha256.digest()).toLowerCase();
+        String sha384Digest = DatatypeConverter.printHexBinary(sha384.digest()).toLowerCase();
+        String sha512Digest = DatatypeConverter.printHexBinary(sha512.digest()).toLowerCase();
+        hexDigests.put("MD5", md5Digest);
+        hexDigests.put("SHA-1", sha1Digest);
+        hexDigests.put("SHA-256", sha256Digest);
+        hexDigests.put("SHA-384", sha384Digest);
+        hexDigests.put("SHA-512", sha512Digest);
+        if (additionalAlgorithm != null) {
+            String extraDigest = DatatypeConverter.printHexBinary(extraAlgo.digest()).toLowerCase();
+            hexDigests.put(additionalAlgorithm, extraDigest);
+        }
+
+        return hexDigests;
+    }
+
+    /**
+     * Moves an object from one location to another if the object does not exist
+     * 
+     * @param source file to move
+     * @param target where to move the file
+     * 
+     * @return fileMoved (boolean) to confirm file is not a duplicate and has been
+     *         moved
+     * @throws IOException       Unable to create parent directory
+     * @throws SecurityException Insufficient permissions to move file
+     */
+    protected boolean move(File source, File target) throws IOException, SecurityException {
+        if (target.exists()) {
+            return false;
+        }
+
+        File destinationDirectory = new File(target.getParent());
+        // Create parent directory if it doesn't exist
+        if (!destinationDirectory.exists()) {
+            Path destinationDirectoryPath = destinationDirectory.toPath();
+            Files.createDirectories(destinationDirectoryPath);
+        }
+
+        // Move file
+        Path sourceFilePath = source.toPath();
+        Path targetFilePath = target.toPath();
+        try {
+            Files.move(sourceFilePath, targetFilePath, StandardCopyOption.ATOMIC_MOVE);
+            return true;
+        } catch (AtomicMoveNotSupportedException amnse) {
+            logHashFileStore.warn(
+                    "HashFileStore.move - StandardCopyOption.ATOMIC_MOVE failed, trying again without CopyOption.");
```

Comment:
We should keep the move atomic. Is there any reason you try a non-atomic move?

Reply:
When I was reviewing the process, the `"AtomicMoveNotSupportedException` can be thrown when trying to move files across different file systems. I wanted to be explicit in catching that and ensure that we move a file... but in hindsight - it is not a good pattern because the operation can potentially be interrupted without the StandardCopyOption.ATOMIC_MOVE (no longer atomic). I have re-thrown the exception instead of re-attempting a move.